**Pre-Lecture**

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [3]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = villagers_df.shape
print(f"Rows: {rows}, Columns: {columns}")


Rows: 391, Columns: 11


2) 
Rows: 391, Columns: 11

Variables - Characteristics that are being measured

Observations - Individual datapoints measuring selected variables

In [4]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Get basic information about the dataset (e.g., column names, data types, non-null values)
print("Dataset Info:")
villagers_df.info()

# Get summary statistics for numerical columns (e.g., mean, std, min, max)
print("\nSummary Statistics for Numerical Columns:")
print(villagers_df.describe())

# Get a summary of all columns, including categorical ones
print("\nSummary Statistics for Categorical Columns:")
print(villagers_df.describe(include=['object']))

# You can also view the first few rows to get a sense of the dataset structure
print("\nFirst 5 Rows of the Dataset:")
print(villagers_df.head())


Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB

Summary Statistics for Numerical Columns:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

Summary Statistics for Categorical Columns:
            id    name gender species birthday

5) 
An attribute, like df.shape is a characteristic or property of an object in python. When it is accessed, it does not affect the object.

A method, like df.describe(), is a function that can be performed on an object. They might take arguments, and after being called, it changes the object to which it corresponds.

**Post-Lecture**

6) 
From ChatGPT:

Summary:
count: Number of non-null entries.
mean: Average value.
std: Standard deviation (spread of data).
min: Smallest value.
25%: Value at the 25th percentile (first quartile).
50%: Median (middle value).
75%: Value at the 75th percentile (third quartile).
max: Largest value.


7) 
    1) df.dropna() 
    Dropping a column temporarily disables it so that the dataframe can be interpreted with the column excluded. 
    This may be preferential when you want to analyze the dataframe once with this column excluded, but not delete it permanentely.

    2) del df['col']
    Deleting a column removes it permanentely from the data set. You may want to do this if a column is missing too many values or is otherwise not useful.
    3) I haven't confirmed this, but I imagine it's because deleting a column may affect the indeces of the other columns in the dataframe.

8) 
    3) 
    I haven't encountered any errors produced by ChatGPT. Looking up information on google the top results tend to also be highly accurate. I believe both are very reliable but due to the nature of chatbots I believe ChatGPT is more likely to be able to automatically spot mistakes that I've made myself, so the end result may be better and you may arrive to successful and functional code faster than if you rely solely on google.

**ChatGPT Summary**

We've discussed various topics related to handling data in pandas, including:

Difference between attributes and methods: Attributes like df.shape provide information without performing calculations, while methods like df.describe() execute operations on the data.

Summary statistics in df.describe(): Key statistics include count, mean, std, min, 25%, 50% (median), 75%, and max, each providing insights into the dataset's distribution.

Handling missing data:

df.dropna(): Removes rows or columns with missing values, either temporarily or permanently.
del df['col']: Permanently deletes a column, typically used when it's mostly missing or irrelevant.
Difference between dropping and permanently deleting a column: Dropping (df.drop()) can either temporarily or permanently remove a column (depending on the inplace parameter), while del df['col'] always permanently deletes the column.

 